interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+	if !exists {
+		return nil, ErrRuleIDNotFound
+	}
+	if index >= len(rdb.rules) {
+		return nil, ErrInternalInconsistency
+	}
+	return rdb.rules[index], nil
+}
+
+// removeRuleWithID removes the rule with the given ID from the rule DB.
+func (rdb *RuleDB) removeRuleWithID(id prompting.IDType) (*Rule, error) {
+	index, exists := rdb.ids[id]
+	if !exists {
+		return nil, ErrRuleIDNotFound
+	}
+	if index >= len(rdb.rules) {
+		return nil, ErrInternalInconsistency
+	}
+	rule := rdb.rules[index]
+	// Remove the rule with the given ID by copying the final rule in rdb.rules
+	// to its index.
+	rdb.rules[index] = rdb.rules[len(rdb.rules)-1]
+	// Record the ID of the moved rule now before truncating, in case the rule
+	// to remove is the moved rule (so nothing was moved).
+	movedID := rdb.rules[index].ID
+	// Truncate rules to remove the final element, which was just copied.
+	rdb.rules = rdb.rules[:len(rdb.rules)-1]
+	// Update the ID-index mapping of the moved rule.
+	rdb.ids[movedID] = index
+	delete(rdb.ids, id)
+	return rule, nil
+}
+
+// permissionDBForUserSnapInterfacePermission returns the permission DB for the
+// given user, snap, interface, and permission.
+func (rdb *RuleDB) permissionDBForUserSnapInterfacePermission(user uint32, snap string, iface string, permission string) *permissionDB {
+	userSnaps := rdb.perUser[user]
+	if userSnaps == nil {
+		userSnaps = &userDB{
+			PerSnap: make(map[string]*snapDB),
+		}
+		rdb.perUser[user] = userSnaps
+	}
+	snapInterfaces := userSnaps.PerSnap[snap]
+	if snapInterfaces == nil {
+		snapInterfaces = &snapDB{
+			PerInterface: make(map[string]*interfaceDB),
+		}
+		userSnaps.PerSnap[snap] = snapInterfaces
+	}
+	interfacePerms := snapInterfaces.PerInterface[iface]
+	if interfacePerms == nil {
+		interfacePerms = &interfaceDB{
+			PerPermission: make(map[string]*permissionDB),
+		}
+		snapInterfaces.PerInterface[iface] = interfacePerms
+	}
+	permVariants := interfacePerms.PerPermission[permission]
+	if permVariants == nil {
+		permVariants = &permissionDB{
+			VariantEntries: make(map[string]variantEntry),
+		}
+		interfacePerms.PerPermission[permission] = permVariants
+	}
+	return permVariants
+}
+
+// RuleConflict stores the rendered variant which conflicted with that of
+// another rule, along with the ID of that conflicting rule.
+type RuleConflict struct {
+	Variant       string           `json:"pattern-variant"`
+	ConflictingID prompting.IDType `json:"conflicting-rule-id"`
+}
+
+// addRulePermissionToTree adds all the path pattern variants for the given
+// rule to the map for the given permission.
+//
+// If there are conflicting pattern variant froms other rules, all variants
+// which were previously added during this function call are removed
+// from the path map, and an error is returned along with a list of the
+// conflicting variants and the IDs of the conflicting rules.
+func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) (error, []RuleConflict) {
```

Comment:
the more usual thing would be to have an error struct that contains the conflicts?

Reply:
I've added an error which contains the conflicts and unwraps to `ErrPathPatternConflict`.